pinot-spi/src/main/java/org/apache/pinot/spi/config/table/IndexConfig.java

```diff
@@ -0,0 +1,46 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.spi.config.table;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.apache.pinot.spi.config.BaseJsonConfig;
+
+
+/**
+ * This is the base class used to configure indexes.
+ *
+ * The common logic between all indexes is that they can be enabled or disabled.
+ *
+ * Indexes that do not require extra configuration can directly use this class.
+ */
+public class IndexConfig extends BaseJsonConfig {
+  public static final IndexConfig ENABLED = new IndexConfig(true);
+  public static final IndexConfig DISABLED = new IndexConfig(false);
+  private final boolean _enabled;
```

Comment:
Probably make it `_disabled` because in most cases the index should be enabled when configured.

Reply:
I don't get why that should be reflected in the name of the attribute. How is it configured in JSON is unrelated to the name of the attribute. In following PRs we will decide what to do with this, but what I initially did was to make it optional, so when an index is specified you usually do not write the `enable`/`disable` attribute unless you actually want to disable it.

I prefer to use `enable` to avoid too many negations and reduce the reading complexity.

Reply:
My concern is that when deserializing from JSON and `enabled` is not explicitly set, the config will have it as `false`. In most of the cases, unless we want to explicitly disable an index, it should be enabled when configured.

Reply:
> My concern is that when deserializing from JSON and enabled is not explicitly set, the config will have it as false.

That is the default behavior of Jackson, but that default behavior shall not modify our code. Another relevant case: IMO subclasses of `IndexConfig` should be immutable. The design of this class enforces it. That means that we need to create a factory method or an annotation annotated with `@JsonCreator`. That method/class have to have one parameter for each property know by Jackson. Therefore we need to introduce the `enabled` (or `disabled`) property there. For example, a `DictionaryIndexConfig` should have a constructor like:

```java
  @JsonCreator
  public DictionaryIndexConfig(@JsonProperty("enabled") boolean enabled, @JsonProperty("onHeap") boolean onHeap,
      @JsonProperty("useVarLengthDictionary") @Nullable Boolean useVarLengthDictionary) {
    super(enabled);
    _onHeap = onHeap;
    _useVarLengthDictionary = Boolean.TRUE.equals(useVarLengthDictionary);
  }
```

Given that we already need to do that. It cost nothing to do:

```java
  @JsonCreator
  public DictionaryIndexConfig(@JsonProperty("enabled") Boolean enabled, @JsonProperty("onHeap") boolean onHeap,
      @JsonProperty("useVarLengthDictionary") @Nullable Boolean useVarLengthDictionary) {
    super(enabled == null || enabled);
    _onHeap = onHeap;
    _useVarLengthDictionary = Boolean.TRUE.equals(useVarLengthDictionary);
  }
```

By doing that tell Jackson what to do without making Jackson defaults have an impact on our code.